WIP

lilic opened the pull request:
@krnowak enjoy 😄 
Lets ignore the number of commits (will squash at the end and make sensible commit msgs) and the fact that I haven't rebased to the latest changes (will do that as well). 😃

krnowak commented on src/yelp-window.c:
```diff
@@ -1158,7 +1165,8 @@ view_uri_selected (YelpView     *view,
     doc_uri = yelp_uri_get_document_uri (uri);
     if (priv->doc_uri == NULL || !g_str_equal (priv->doc_uri, doc_uri)) {
-        window_set_bookmarks (window, doc_uri);
+        if (doc_uri)
```
Do we still need this?

lilic replied:
Hmm checked, seems to work fine without it. Guess not.

krnowak commented on src/yelp-window.c:
```diff
@@ -198,6 +198,13 @@ yelp_window_init (YelpWindow *window)
 }
 
+YelpView *
+yelp_window_get_view (YelpWindow *window){
```
We can drop this function - we don't use it anymore.

krnowak commented on src/yelp-application.c:
```diff
@@ -162,6 +204,12 @@ yelp_application_init (YelpApplication *app)
 static void
+yelp_application_activate (GApplication *application)
+{
+
```
Maybe add a TODO about showing the window here if an app already has any window?

krnowak commented on src/yelp-application.c:
```diff
@@ -117,6 +117,48 @@ struct _YelpApplicationPrivate {
+{
+    YelpApplication *app = YELP_APPLICATION (user_data);
+    YelpUri *uri = yelp_uri_new (DEFAULT_URI);
+    YelpUri *page_uri = yelp_uri_new_search (uri, g_variant_get_string(parameter, NULL));
```
You should unref this variable at the end of the function.

krnowak commented on src/yelp-application.c:
```diff
@@ -117,6 +117,48 @@ struct _YelpApplicationPrivate {
+}
+
+static void
+activate_show_search(GSimpleAction *action,
```
Coding style: space between the function name and the opening parenthesis. Align the parameters to the opening parenthesis properly.

krnowak commented on src/yelp-application.c:
```diff
@@ -117,6 +117,48 @@ struct _YelpApplicationPrivate {
+          gboolean         fallback_help_list);
+
+static void 
+activate_show_page(GSimpleAction *action,
```
Coding style: space between the function name and the opening parenthesis. Align the parameters to the opening parenthesis properly.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+#include <gio/gio.h>
+#include <string.h>
+#include <glib/gi18n.h>
+#include <gdk/gdkx.h>
```
Drop this include. It's unused.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+
+#include <gio/gio.h>
+#include <string.h>
+#include <glib/gi18n.h>
```
Do we need this?

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+
+struct _YelpSearchProviderApp {
+    GApplication parent;
+    gulong id;
```
Maybe call it `uri_resolve_id`?

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+    GDBusObjectManagerServer *object_manager;
+    YelpShellSearchProvider2 *skeleton;
+
+    YelpDocument *doc;
```
We probably do not need this anymore, right?

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+                  gpointer user_data)
+{
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+    GVariantBuilder meta, metas;
```
I would move the `meta` variable inside the `for` loop.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+{
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+    GVariantBuilder meta, metas;
+    gchar *gicon_str;
```
I would move the `gicon_str` variable inside the `for` loop.

LiliC replied:
@krnowak And then also free it in the for loop?

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+    return ret;
+}
+
+static GVariantBuilder
```
Ha, that's unusual. :) I haven't seen anyone returning `GVariantBuilder` by value before. Anyway, to avoid complaints from upstream or other reviewers I would rather recommend returning a `(as)` typed `GVariant *`, so at the end of the function, instead of `return builder`, do `return g_variant_new ("(as)", &builder);`. So at the call sites you can pass the returned GVariant directly to `g_dbus_method_invocation_return_value`.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+    GVariantBuilder builder;
+    YelpSearchProviderApp *app = YELP_SEARCH_PROVIDER_APP (user_data);
+    GHashTableIter iter;
+    GPtrArray *res = g_ptr_array_new();
```
Space between function name and opening parentheses.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+
+static GVariantBuilder
+get_search_results (gchar* term,
+                    gpointer user_data)
```
It would be better if instead of having `gpointer user_data`, we could have just `GHashTable *page_data` because:
- it would be more obvious what we pass here if we specified `YelpSearchProviderApp *app`
- `app->page_data` is the only thing we need from `YelpSearchProviderApp`, so let's pass the hash map directly instead.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+                               gchar **terms,
+                               gpointer user_data)
+{
+    gchar *joined_terms = g_strjoinv (" ", terms);
```
Just declare the variable here, but do the `g_strjoinv` after the `if (…)` to easily avoid doing useless work in case of early return. Also, this would avoid a memory leak in the early return case.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+{
+    gchar *joined_terms = g_strjoinv (" ", terms);
+
+    if (g_strv_length (terms) < 2 && g_utf8_strlen (terms[0], -1) < 3 ){
```
Hm hm hm, makes me wonder if this is a good reason to avoid searching, when some Chinese writes a single ideogram that means a printer and nothing will be searched. Something for the upstream to decide I guess…

lilic replied:
Have seen this other search providers, where they avoid searching 1 or 2 or 3 char. But yes its up to them, I am fine with removing it.

krnowak replied:
Let's keep it for now, we can ask Cosimo or upstream about it.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+        g_dbus_method_invocation_return_value (invocation, g_variant_new ("(as)", NULL));
+        return;
+    }
+    GVariantBuilder builder = get_search_results(joined_terms, user_data);
```
Space between function name and opening parenthesis. Also, variable declaration should go to the top of the file, otherwise the compiler will complain.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+                                 gchar **terms,
+                                 gpointer user_data)
+{
+    gchar *joined_terms = g_strjoinv (" ", terms);
```
Just declare the variable here, but do the `g_strjoinv` after the `if (…)` to easily avoid doing useless work in case of early return. Also, this would avoid a memory leak in the early return case.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+{
+    gchar *joined_terms = g_strjoinv (" ", terms);
+
+    if (g_strv_length (terms) < 2 && g_utf8_strlen (terms[0], -1) < 3 ){
```
Hm hm hm, makes me wonder if this is a good reason to avoid searching, when some Chinese writes a single ideogram that means a printer and nothing will be searched. Something for the upstream to decide I guess…

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+        g_dbus_method_invocation_return_value (invocation, g_variant_new ("(as)", NULL));
+        return;
+    }
+    GVariantBuilder builder = get_search_results(joined_terms, user_data);
```
Space between function name and opening parenthesis. Also, variable declaration should go to the top of the file, otherwise the compiler will complain.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+
+static GVariant *
+get_result_metas (gchar **page_ids,
+                  gpointer user_data)
```
It would be better if instead of having `gpointer user_data`, we could have just `GHashTable *page_data` because:
- it would be more obvious what we pass here if we specified `YelpSearchProviderApp *app`
- `app->page_data` is the only thing we need from `YelpSearchProviderApp`, so let's pass the hash map directly instead.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+                         gpointer user_data)
+{
+    g_dbus_method_invocation_return_value (invocation,
+                                          get_result_metas (results, user_data));
```
Indentation.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+
+    g_variant_builder_init (&param_builder, G_VARIANT_TYPE ("av"));
+    for (i = 0; terms[i] != NULL; i++)
+        g_variant_builder_add (&param_builder, "v", g_variant_new_string (joined_terms));
```
If you search for `foo bar baz` then you append the "foo bar baz" thrice here. I think you wanted to pass `terms[i]` instead of `joined_terms` here. Also, this would make `joined_terms` variable unused, so you could remove it.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+                        guint32                             timestamp,
+                        gpointer                            user_data)
+{
+    GDBusConnection *session_bus;
```
Here you seem to do similar steps as in the `handle_launch_search` function. The only difference is the action name and the contents of the GVariantBuilder. How about refactoring this a bit to avoid repetition?

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+    g_signal_connect (self->skeleton, "handle-launch-search",
+                    G_CALLBACK (handle_launch_search), self);
+
+    g_dbus_interface_skeleton_export (G_DBUS_INTERFACE_SKELETON (self->skeleton),
```
I would pass a GError instance here instead of `NULL` and check if this function returns FALSE. If so, I would add some g_warning about the fact that we failed to export the interface on the connection.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+    YelpSearchProviderApp *self = YELP_SEARCH_PROVIDER_APP (app);
+    gchar **page_ids = yelp_document_list_page_ids (document);
+    gchar **iter;
+
```
I think that in general you should handle the `YelpDocumentSignal signal` parameter here. If it is:
- `YELP_DOCUMENT_SIGNAL_ERROR` - print a warning (print `error->message` too) and jump to g_bus_own_name.
- `YELP_DOCUMENT_SIGNAL_CONTENTS` then probably return early, because we don't care about it at all
- `YELP_DOCUMENT_SIGNAL_INFO` do what we do now.

lilic replied:
I have done this, just github doesn't hide it...

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+                             "id", g_variant_new_string (page_ids[i]));
+      g_variant_builder_add (&meta, "{sv}",
+                             "name", g_variant_new_string (data->title));
+      gicon_str = g_icon_to_string (data->icon);
```
I'd do `gicon_str = (data->icon != NULL) ? g_icon_to_string (data->icon) : NULL` to avoid criticals when during preloading we got some invalid icon.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+        gchar *page_id = *iter;
+        PageData *data = g_new0 (PageData, 1);
+        gchar *icon_string = yelp_document_get_page_icon (document, page_id);
+        GIcon *icon = g_icon_new_for_string (icon_string, &error);
```
I would just pass NULL here I think, because you do not do anything with the error. But if you want to print an error (with g_warning or g_debug) then you need to pass your own instance of a GError here, because the `error` variable comes from libyelp and is meant to be used for reporting an error from libyelp.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+        g_free (icon_string);
+
+        data = page_data_new_steal (yelp_document_get_page_title (document, page_id), 
+                                yelp_document_get_page_desc (document, page_id), 
```
Indentation.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+                                         search_provider_name_acquired_cb,
+                                         search_provider_name_lost_cb,
+                                         app, NULL);
+    g_free (page_ids);
```
Maybe you could add a comment here that you stole the contents of `page_ids` so the only thing you free is the array that held the pointers to contents.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+    if (g_getenv ("YELP_SEARCH_PROVIDER_PERSIST") != NULL)
+        g_application_hold (app);
+
+    base_uri = yelp_uri_new ("help:gnome-help");
```
I think you should unref the `base_uri` after calling `yelp_uri_resolve` at the end of the function - `yelp_uri_resolve` will take a ref on this variable on its own.

krnowak replied:
Also, how about adding something like `#define YELP_URI_DEFAULT "help:gnome-help"` somewhere in `yelp-uri.h`, so you could use it in both `yelp-application.c` and here?

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+                               app,
+                               NULL);
+
+    g_object_unref (uri);
```
This unref should not be needed I think.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+    gchar *page_id;
+
+    doc = yelp_document_get_for_uri (uri);
+    g_signal_handler_disconnect (uri, app->id);
```
You should set `app->id` to zero afterwards.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+yelp_search_provider_app_init (YelpSearchProviderApp *self)
+{
+    g_application_set_inactivity_timeout (self, SEARCH_PROVIDER_INACTIVITY_TIMEOUT);
+    g_application_hold(self);
```
This should be dropped I think…

krnowak replied:
Or, add `g_application_hold` in `search_provider_app_startup`, before calling `yelp_uri_resolve` and then call `g_application_release` after `g_bus_own_name` in `preload_data_cb`

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+{
+    GApplication *app;
+    gint res;
+    gchar *env_var = "YELP_SEARCH_PROVIDER";
```
Is this still needed? I think that calling `gtk_init` below fixed the issue with settings stuff, no?

lilic replied:
No, I tested it, we still need it.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+    gtk_init (&argc, &argv);
+    gboolean suc = g_setenv (env_var, "search", TRUE);
+    if (suc == FALSE)
+        g_print ("We failed setting out env variable %s\n", env_var);
```
g_warning.

krnowak commented on libyelp/yelp-search-entry.c:
```diff
@@ -466,7 +466,6 @@ entry_match_func (GtkEntryCompletion *completion,
     gtk_tree_model_get (model, iter, COMPLETION_COL_FLAGS, &flags, -1);
     if (flags & COMPLETION_FLAG_ACTIVATE_SEARCH)
         return TRUE;
-
```
Drop this change.

krnowak commented on libyelp/yelp-settings.c:
```diff
@@ -599,11 +599,18 @@ yelp_settings_get_default (void)
-				 "gtk-settings", gtk_settings_get_default (),
-				 "gtk-icon-theme", gtk_icon_theme_get_default (),
-				 NULL);
+    if (settings == NULL) {
```
Not sure if this is still needed provided that we call gtk_init… Could you test it?

lilic replied:
The `settings == NULL` check was there before.

Checked the rest, we still need it despite the gtk_init.

lilic replied:
Removed gtk_init, but kept this.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+
+#include "libyelp/yelp-document.h"
+#include "libyelp/yelp-uri.h"
+#include "libyelp/yelp-view.h"
```
We can drop it, we don't use `YelpView` here.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+
+    g_variant_builder_init (&metas, G_VARIANT_TYPE ("aa{sv}"));
+
+    for (i = 0; page_ids[i] != NULL; i++){
```
Space between closing parenthesis and opening bracket.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+    g_variant_builder_init (&metas, G_VARIANT_TYPE ("aa{sv}"));
+
+    for (i = 0; page_ids[i] != NULL; i++){
+      PageData *data = g_hash_table_lookup (app->page_data_hash_map, page_ids[i]);
```
This block should be indented with 2 more spaces.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+get_search_results (gchar* term,
+                    gpointer user_data)
+{
+
```
Drop this empty line.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+}
+
+static void
+handle_get_result_metas (YelpShellSearchProvider2  *skeleton,
```
The parameter names are oddly spaced here.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+}
+
+static void
+handle_launch_search (YelpShellSearchProvider2  *skeleton,
```
The parameter names are oddly spaced here.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+}
+
+static void
+handle_activate_result (YelpShellSearchProvider2  *skeleton,
```
The parameter names are oddly spaced here.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+
+static void
+preload_data_cb (YelpDocument          *document,
+                       YelpDocumentSignal    signal,
```
Indentation.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+    gchar **page_ids = yelp_document_list_page_ids (document);
+    gchar **iter;
+
+    for (iter = page_ids; *iter; iter++){
```
Space between closing parenthesis and opening bracket.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+    }
+
+    self->name_owner_id = g_bus_own_name (G_BUS_TYPE_SESSION,
+                                         "org.gnome.Yelp.SearchProvider",
```
Indentation.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+    g_signal_handler_disconnect (uri, app->id);
+    page_id = yelp_uri_get_page_id (uri);
+    yelp_document_request_page (doc,
+                               page_id,
```
Indentation.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,498 @@
+    YelpUri *base_uri;
+    G_APPLICATION_CLASS (yelp_search_provider_app_parent_class)->startup (app);
+
+    // TODO: remove holding the app forever
```
I'd keep this. Useful for debugging.

lilic replied:
Keep it? You mean for now, or keep it for later as well?

krnowak replied:
For later as well - it is hidden behind the env var, so it can't be triggered easily.

krnowak commented on src/yelp-application.c:
```diff
@@ -224,7 +273,6 @@ yelp_application_finalize (GObject *object)
     G_OBJECT_CLASS (yelp_application_parent_class)->finalize (object);
 }
 
-
```
Drop this change.

krnowak commented on src/yelp-application.c:
```diff
@@ -246,9 +294,11 @@ yelp_application_cmdline (GApplication     *app,
-            newuri = g_file_get_uri (new);
-            g_free ((*arguments)[i]);
-            (*arguments)[i] = newuri;
+            if (g_file_query_exists (new, NULL)) {
```
Maybe add a comment about it being a hack for handling the `--gapplication-service` flag which is unlikely to exist as a file in the current working directory?

krnowak replied:
A terrible terrible hack, for that matter.

krnowak commented on src/yelp-window.h:
```diff
@@ -53,4 +53,5 @@ void              yelp_window_get_geometry (YelpWindow      *window,
                                             gint            *width,
                                             gint            *height);
 
+YelpView *        yelp_window_get_view (YelpWindow *window);
```
Can be dropped.

LiliC commented:
@krnowak Can you have another look please? Thanks!

krnowak commented on libyelp/yelp-settings.c:
```diff
@@ -599,11 +599,18 @@ yelp_settings_get_default (void)
-				 "gtk-icon-theme", gtk_icon_theme_get_default (),
-				 NULL);
+    if (settings == NULL) {
+        const gchar * yelp_search_provider = g_getenv("YELP_SEARCH_PROVIDER");
```
Could you rename this env var to something more descriptive? Like `YELP_SKIP_GUI`? Same with the variable name.

krnowak commented on libyelp/yelp-uri.c:
```diff
@@ -244,6 +244,7 @@ yelp_uri_resolve (YelpUri *uri)
         g_signal_connect_swapped (priv->res_base, "resolved",
                                   G_CALLBACK (resolve_start),
                                   uri);
+        g_object_ref (uri);
```
I'm a bit on the fence with this change. I mean - keep it please, make sure it is in a separate commit, but this may need some investigation later whether this `g_object_ref` needs some `g_object_unref` counterpart somewhere, otherwise we may be leaking a reference, dunno.

krnowak commented on po/POTFILES.in:
```diff
@@ -21,7 +21,7 @@ libyelp/yelp-storage.c
 libyelp/yelp-transform.c
 libyelp/yelp-uri.c
 libyelp/yelp-view.c
+yelp.desktop.in.in
```
You probably want to drop all the changes in this file.

krnowak commented on po/POTFILES.skip:
```diff
@@ -1,9 +1,9 @@
+yelp.desktop.in
```
You probably want to drop all the changes in this file.

krnowak commented on search-provider/Makefile.am:
```diff
@@ -0,0 +1,40 @@
+AM_CPPFLAGS = \
```
Instead of using `AM_CPPFLAGS` I would use `yelp_search_provider_CFLAGS`.

krnowak commented on search-provider/Makefile.am:
```diff
@@ -0,0 +1,40 @@
+AM_CPPFLAGS = \
+	$(YELP_CFLAGS) \
+	-DDIRECTORY_DIR=\"$(directorydir)\" \
```
gitgrepping for `DIRECTORY_DIR` and for `GNOMELOCALEDIR` didn't return me anything meaningful, so you can just drop these lines.

lilic replied:
so just keeping `YELP_CFLAGS` line?

krnowak replied:
Yes, that would be the only thing here.

krnowak commented on search-provider/Makefile.am:
```diff
@@ -0,0 +1,40 @@
+	-DGNOMELOCALEDIR=\"$(datadir)/locale\" \
+	$(NULL)
+
+bin_PROGRAMS = yelp-search-provider
```
I don't think that `yelp-search-provider` should end up in `/usr/bin`. I think it should be placed in `/usr/libexec` or something. So instead of `bin_PROGRAMS` use `libexec_PROGRAMS`.

krnowak commented on search-provider/Makefile.am:
```diff
@@ -0,0 +1,40 @@
+service_in_files = org.gnome.Yelp.SearchProvider.service.in
+
+org.gnome.Yelp.SearchProvider.service: org.gnome.Yelp.SearchProvider.service.in Makefile
+	$(AM_V_GEN) sed -e "s|\@bindir\@|$(bindir)|" $< > $@
```
These should be `libexecdir` instead of `bindir`.

krnowak commented on search-provider/Makefile.am:
```diff
@@ -0,0 +1,40 @@
+org.gnome.Yelp.SearchProvider.service: org.gnome.Yelp.SearchProvider.service.in Makefile
+	$(AM_V_GEN) sed -e "s|\@bindir\@|$(bindir)|" $< > $@
+
+BUILT_SOURCES =	$(dbus_shell_search_provider_built_sources)
```
Drop the extra spaces here.

krnowak commented on search-provider/org.gnome.Yelp.SearchProvider.service.in:
```diff
@@ -0,0 +1,3 @@
+[D-BUS Service]
+Name=org.gnome.Yelp.SearchProvider
+Exec=@bindir@/yelp-search-provider
```
This would need to be `libexecdir` instead of `bindir`.

krnowak commented on src/yelp-application.c:
```diff
@@ -117,6 +117,49 @@ struct _YelpApplicationPrivate {
+
+    open_uri (app, page_uri, TRUE, FALSE);
+
+    g_object_unref(uri);
```
Space between function name and opening parenthesis.

krnowak commented on src/yelp-application.c:
```diff
@@ -117,6 +117,49 @@ struct _YelpApplicationPrivate {
+    open_uri (app, page_uri, TRUE, FALSE);
+
+    g_object_unref(uri);
+    g_object_unref(page_uri);
```
Space between function name and opening parenthesis.

krnowak commented on src/yelp-application.c:
```diff
@@ -117,6 +117,49 @@ struct _YelpApplicationPrivate {
+
+    g_object_unref(uri);
+    g_object_unref(page_uri);
+    g_free(xref);
```
Space between function name and opening parenthesis.

krnowak commented on src/yelp-application.c:
```diff
@@ -117,6 +117,49 @@ struct _YelpApplicationPrivate {
+
+    open_uri (app, page_uri, TRUE, FALSE);
+
+    g_object_unref(uri);
```
Space between function name and opening parenthesis.

lilic replied:
I really miss those always :/

krnowak replied:
You just wait and see - you'll start adding those spaces in another project where they don't want those spaces. ;)

krnowak commented on src/yelp-application.c:
```diff
@@ -117,6 +117,49 @@ struct _YelpApplicationPrivate {
+    open_uri (app, page_uri, TRUE, FALSE);
+
+    g_object_unref(uri);
+    g_object_unref(page_uri);
```
Space between function name and opening parenthesis.

krnowak commented on src/yelp-application.c:
```diff
@@ -162,6 +205,12 @@ yelp_application_init (YelpApplication *app)
 static void
+yelp_application_activate (GApplication *application)
+{
+    // TODO: Show the window if an app already has it.
```
Just for the record - does pressing yelp icon somewhere in gnome shell bring up the yelp window with this function being empty?

krnowak replied:
If so, then maybe this TODO is useless, and rather a comment about no necessary steps here would be better…

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+}
+
+static GVariant *
+get_result_metas (gchar     **page_ids,
```
One more space between the type and the asterisk, please?

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+        g_variant_builder_init (&meta, G_VARIANT_TYPE ("a{sv}"));
+        g_variant_builder_add (&meta, "{sv}",
+                              "id", g_variant_new_string (page_ids[i]));
+        g_variant_builder_add (&meta, "{sv}",
```
I think that at this point we should check if the `data` variable is not null (could be when somebody sends us some rubbish over D-Bus). So something like:
```
if (data != NULL) {
    g_variant_builder_add (…);
    …;
}
else {
    g_warning ("Invalid page ID %s", page_ids[i]);
    // fill the meta builder with "invalid" as a "name" and "description"?
}
g_variant_builder_add_value (&metas, …);
```

lilic replied:
Wouldn't it be better to just return `NULL` name and desc instead of filling with "invalid"?

krnowak replied:
That's fine too I guess.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+}
+
+
+static gboolean
```
Maybe add a `FIXME` saying that this was copied from yelp-search-entry (IIRC) and adapted, so maybe it should be exported in libyelp too to avoid repetition.

lilic replied:
Agreed. I think that will need to adjusted, agreed  and exported so that one search is used across yelp first, then we can use it as well in the search provider.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+}
+
+static GVariant *
+get_search_results (gchar     **terms,
```
One more space between the type and the asterisk, please?

lilic replied:
Hmm not sure I agree :p 
I aligned to the name here... [Example from yelp](https://github.com/kinvolk/yelp/blob/master/src/yelp-application.c#L48-L51)

krnowak replied:
Counter example - https://github.com/kinvolk/yelp/blob/master/src/yelp-application.c#L229

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+    return g_variant_new ("(as)", &results);
+}
+
+typedef GVariant * (*ResultGetter) (gchar     **terms,
```
One more space between the type and the asterisk, please?

lilic replied:
Hmm not sure I agree :p 
I aligned to the name here... [Example from yelp](https://github.com/kinvolk/yelp/blob/master/src/yelp-application.c#L48-L51)

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+{
+    YelpSearchProviderApp *app;
+    GDBusMethodInvocation *invocation;
+    gchar                **terms;
```
One more space between the type and the asterisk, please?

lilic replied:
Hmm not sure I agree :p 
I aligned to the name here... [Example from yelp](https://github.com/kinvolk/yelp/blob/master/src/yelp-application.c#L48-L51)

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+static DelayedInitialResultSet *
+delayed_initial_result_set_new (YelpSearchProviderApp *app,
+                                GDBusMethodInvocation *invocation,
+                                gchar                **terms,
```
One more space between the type and the asterisk, please?

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+static gboolean 
+handle_get_initial_result_set (YelpShellSearchProvider2 *skeleton,
+                               GDBusMethodInvocation    *invocation,
+                               gchar                   **terms,
```
One more space between the type and the asterisk, please?

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+static gboolean
+handle_get_subsearch_result_set (YelpShellSearchProvider2 *skeleton,
+                                 GDBusMethodInvocation    *invocation,
+                                 gchar                   **previous_results,
```
One more space between the type and the asterisk, please?

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+yelp_search_provider_app_dbus_register (GApplication    *app,
+                                        GDBusConnection *connection,
+                                        const gchar     *object_path,
+                                        GError         **error)
```
One more space between the type and the asterisk, please?

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+}
+
+int
+main (int argc,
```
A few more spaces here to match `argv`.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+    GApplication *app;
+    gint res;
+    gboolean suc;
+    gchar *env_var = "YELP_SEARCH_PROVIDER";
```
If you change the env var name, change it here too.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+    gboolean suc;
+    gchar *env_var = "YELP_SEARCH_PROVIDER";
+
+    suc = g_setenv (env_var, "search", TRUE);
```
If you change the env var name, then maybe `"yes"` instead of `"search"` would be a better value.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+
+    suc = g_setenv (env_var, "search", TRUE);
+    if (suc == FALSE)
+        g_warning ("We failed setting out env variable %s\n", env_var);
```
`"failed to set the env var %s"`?

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+    GVariantBuilder results;
+    gchar *term = g_strjoinv (" ", terms);
+    GHashTableIter iter;
+    GPtrArray *res = g_ptr_array_new ();
```
Drop this variable, seems to be unused. And is leaking.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+    gchar *term = g_strjoinv (" ", terms);
+    GHashTableIter iter;
+    GPtrArray *res = g_ptr_array_new ();
+    gint i;
```
Drop this variable too - it is unused.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+    GDBusMethodInvocation *invocation;
+    gchar                **terms;
+    ResultGetter           result_getter;
+} DelayedInitialResultSet;
```
Consider renaming it to `DelayedResultGetter` or something like this. It is not only about delayed getting of the initial results set.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+{
+    DelayedInitialResultSet* delayed = user_data;
+
+    if (delayed->app->page_data_hash_map != NULL) {
```
Ah, this is always true (we set this pointer n the startup). So we should:
- either change this to `g_hash_table_size (delayed->app->page_data_hash_map) > 0`
- or move `g_hash_table_new_full` from `search_provider_app_startup` to `preload_data_cb` inside the `YELP_DOCUMENT_SIGNAL_INFO` block.

krnowak replied:
I think that the first solution is much less work.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+}
+
+static gboolean
+yelp_launch (gchar                 *terms,
```
I'd rename `terms` to `string_parameter`.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+                                         NULL,
+                                         &error);
+
+    g_object_unref (session_bus);
```
This should not be unreffed - `g_dbus_interface_skeleton_get_connection` does not add a reference to the returned `GDBusConnection` instance (it has `transfer none` annotation).

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+    return TRUE;
+}
+
+static gboolean
```
This function always returns TRUE, so I would just make it `static void`.

lilic replied:
I did github, I did do this comment... :/

krnowak replied:
Yeah, github sometimes moves the comment elsewhere, to the fitting location. It did it to me too.

krnowak replied:
Or maybe not - the line numbers indicate it is still at yelp_launch. Well, bugs happen.

krnowak replied:
Ah, now I see that github had indeed moved it. The line below says `handle_get_subsearch_result_set`.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+                      guint32                    timestamp,
+                      gpointer                   user_data)
+{
+    gchar *joined_terms = g_strjoinv (" ", terms);
```
You are leaking the `joined_terms` variable here. You need to free it, before returning from the function.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+
+static gboolean
+yelp_launch (gchar                 *terms,
+             gpointer               user_data,
```
I would prefer if this function was not taking any `gpointer user_data` but rather `YelpSearchProviderApp *app` instead. That way it is clear what is really passed.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+                      gpointer                   user_data)
+{
+    gchar *joined_terms = g_strjoinv (" ", terms);
+    return yelp_launch (joined_terms, user_data, invocation, "show-search");
```
Instead of `user_data` I'd prefer if you passed `YELP_SEARCH_PROVIDER_APP (user_data)` instead - more typing, but clearer for the reader. Also, if `yelp_launch` has a `void` return type, you'll need to return TRUE explicitly.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+                        guint32                    timestamp,
+                        gpointer                   user_data)
+{
+    return yelp_launch (result, user_data, invocation, "show-page");
```
Instead of `user_data` I'd prefer if you passed `YELP_SEARCH_PROVIDER_APP (user_data)` instead - more typing, but clearer for the reader. Also, if `yelp_launch` has a `void` return type, you'll need to return TRUE explicitly.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+    gchar **iter;
+
+    if (signal == YELP_DOCUMENT_SIGNAL_ERROR) {
+        if (error)
```
I'd do `g_warning ("error during preloading data: %s", (error != NULL) ? error->message : "unknown error")`

lilic replied:
I did this one as well.

krnowak replied:
Ah, but the `if (error)` is not needed anymore, right?

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+        }
+
+        g_application_release (G_APPLICATION (self));
+        g_free (page_ids);
```
I think I asked for adding a comment here about stealing the contents of the `page_ids` array, so we free the array only here with `g_free` instead of the whole thing with `g_strfreev`.

lilic replied:
Must have missed that.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+    }
+    g_dbus_method_invocation_return_value (invocation, NULL);
+    return TRUE;
+
```
Please remove this empty line.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+    if (g_getenv ("YELP_SEARCH_PROVIDER_PERSIST") != NULL)
+        g_application_hold (app);
+
+    base_uri = yelp_uri_new ("help:gnome-help");
```
What about my earlier suggestion about exporting the `DEFAULT_URI` macro in yelp-uri.h as `YELP_GNOME_HELP_URI`, so you could use it here?

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+    self->skeleton = yelp_shell_search_provider2_skeleton_new ();
+
+    g_signal_connect (self->skeleton, "handle-get-initial-result-set",
+                    G_CALLBACK (handle_get_initial_result_set), self);
```
Indentation.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+    g_signal_connect (self->skeleton, "handle-get-initial-result-set",
+                    G_CALLBACK (handle_get_initial_result_set), self);
+    g_signal_connect (self->skeleton, "handle-get-subsearch-result-set",
+                    G_CALLBACK (handle_get_subsearch_result_set), self);
```
Indentation.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+    g_signal_connect (self->skeleton, "handle-get-subsearch-result-set",
+                    G_CALLBACK (handle_get_subsearch_result_set), self);
+    g_signal_connect (self->skeleton, "handle-get-result-metas",
+                    G_CALLBACK (handle_get_result_metas), self);
```
Indentation.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+    g_signal_connect (self->skeleton, "handle-get-result-metas",
+                    G_CALLBACK (handle_get_result_metas), self);
+    g_signal_connect (self->skeleton, "handle-activate-result",
+                    G_CALLBACK (handle_activate_result), self);
```
Indentation.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+    g_signal_connect (self->skeleton, "handle-activate-result",
+                    G_CALLBACK (handle_activate_result), self);
+    g_signal_connect (self->skeleton, "handle-launch-search",
+                    G_CALLBACK (handle_launch_search), self);
```
Indentation.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,568 @@
+{
+    GApplication *app;
+    gint res;
+    gboolean suc;
```
I would drop this `suc` variable and inline the `g_setenv` inside the `if`:
```
if (!g_setenv (…)) {
    …
}
```

krnowak commented on src/yelp-application.c:
```diff
@@ -117,6 +117,49 @@ struct _YelpApplicationPrivate {
+                    gpointer       user_data)
+{
+    YelpApplication *app = YELP_APPLICATION (user_data);
+    YelpUri *uri = yelp_uri_new (DEFAULT_URI);
```
What about my earlier suggestion about exporting the `DEFAULT_URI` macro in yelp-uri.h as `YELP_GNOME_HELP_URI`?

lilic replied:
hmm what if we define that macro in yelp-application instead? As we do not use it anywhere in yelp-uri.

krnowak replied:
YelpApplication is part of an yelp application (in the `src` directory), not of the libyelp, right?

krnowak commented on src/yelp-application.c:
```diff
@@ -117,6 +117,49 @@ struct _YelpApplicationPrivate {
+                      gpointer       user_data)
+{
+    YelpApplication *app = YELP_APPLICATION (user_data);
+    YelpUri *uri = yelp_uri_new (DEFAULT_URI);
```
Ditto about exporting the `DEFAULT_URI` macro.

krnowak reviewed (changes requested):
Looks good. Don't be scared with the number of comments - most of them are just stylistic nitpicks.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,573 @@
+
+
+static DelayedResultGetter *
+delayed_initial_result_set_new (YelpSearchProviderApp  *app,
```
If you renamed `DelayedInitialResultSet` to `DelayedResultGetter`, then you could also rename `delayed_initial_result_set_{new,free}` to `delayed_result_getter_{new,free}`.

krnowak commented on Makefile.am:
```diff
@@ -1,6 +1,6 @@
 ACLOCAL_AMFLAGS = -I m4 ${ACLOCAL_FLAGS}
 
-SUBDIRS = po
+SUBDIRS = . po search-provider
```
Forgot to mention it earlier, but could you make `po` last?

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,581 @@
+            g_variant_builder_add (&meta, "{sv}",
+                                   "name", NULL);
+            g_variant_builder_add (&meta, "{sv}",
+                                  "gicon", NULL);
```
Indentation.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,581 @@
+    return g_variant_new ("(aa{sv})", &metas);
+}
+
+// FIXME: Search was copied from yelp-search-entry and slightly adapted.
```
Please use the C-style comments, for example:
```
/* FIXME: Search was copied from yelp-search-entry and slightly adapted.
 * When search in yelp gets adjusted we should export it and use it here.
 */
```

krnowak replied:
These are C++ style comments, some pickier options to the compiler may cause it to bail out.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,581 @@
+        }
+
+        g_application_release (G_APPLICATION (self));
+        // We only steal the contents of the page_ids array,
```
C-style comments here, please.

krnowak commented on src/yelp-application.c:
```diff
@@ -162,6 +205,12 @@ yelp_application_init (YelpApplication *app)
 static void
+yelp_application_activate (GApplication *application)
+{
+    // Here we don't need to do anything else.
```
C-style comment here, please.

krnowak reviewed (changes requested):
A few more nitpicks.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,583 @@
+#include <config.h>
```
The license blurb is missing. Just paste the following in the beginning of the file:
```
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
/*
 * Copyright 2017 Endless Mobile, Inc.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public
 * License along with this program; if not, see <http://www.gnu.org/licenses/>.
 *
 * Author: Lili Cosic <lili@kinvolk.io>
 */
```
(I hope I got your name and surname the way you want them to be presented.)

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,583 @@
+            g_variant_builder_add (&results, "s", page_id);
+    }
+
+    g_free(term);
```
Heh heh heh, you know what. ;)

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,583 @@
+
+
+static DelayedResultGetter *
+delayed_initial_result_getter_new (YelpSearchProviderApp  *app,
```
You forgot to remove the `_initial` part from the name.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,583 @@
+}
+
+static void
+delayed_initial_result_getter_free (DelayedResultGetter *delayed)
```
You forgot to remove the `_initial` part from the name.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,583 @@
+    }
+
+    delayed = delayed_initial_result_getter_new (app,
+                                              invocation,
```
After you fixed the name of the function, fix the indentation too.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,583 @@
+    }
+
+    delayed = delayed_initial_result_getter_new (app,
+                                              invocation,
```
After you fixed the name of the function, fix the indentation too.

krnowak commented on search-provider/yelp-search-provider.c:
```diff
@@ -0,0 +1,583 @@
+    }
+
+    delayed = delayed_initial_result_getter_new (app,
+                                              invocation,
```
After you fixed the name of the function, fix the indentation too.

krnowak reviewed (changes requested):
Almost there!

krnowak commented:
Also, what about the changes in the `po` directory?

Commit:
search-provider: add search provider

Commit:
libyelp: add env variable YELP_SKIP_GUI

This adds an option to projects that want to use libyelp but have no GUI,
so they may want to skip some GUI parts libyelp is trying to initialize
and aborts because of some failure.
We are using this env var in yelp search provider.

Commit:
*: make application D-Bus launchable

Changing the application to be launchable from D-Bus,
we needed to rename desktop file according to yelp application ID.
Also add the yelp service file and add --gapplication-service flag to it.

Commit:
po: rename desktop file

Commit:
libyelp: keep parent URI alive

When resolving base URI, keep parent URI alive.

Commit:
*: export default uri as YELP_GNOME_HELP_URI

To avoid duplication, we export the default uri so it can be used
across the project.

Commit:
src: handle --gapplication-service flag

When we launch the application from D-Bus, we just pass in the
--gapplication-service flag, where as yelp expects a file to be passed
in as an argument. So we work around that by checking if the argument we pass is a file.

Commit:
src: add actions to yelp

To be able to launch yelp via the search provider launch_search and
activate_result functions, adjustments had to be made to yelp application.

"show-page" action is used to display specific page.
"show-search" action to display search page with results for given term.

Commit:
gitignore: update file

krnowak commented:
So there are nine commits that you could group into 3 parts:
1. Hacks, workarounds and other preparations to have search provider working (in any order):
    - unref
    - YELP_SKIP_GUI
    - YELP_GNOME_HELP_URI
    - `--gapplication-service` flag hack (should come before the D-Bus activatable commit, though)
    - D-Bus activatable Yelp
    - actions in Yelp 
2. Search provider
    - just one commit that adds the code
3. Random nuisance (in any order you want):
    - po
    - .gitignore

About commit messages - in general the first line usually is in format `<project part>: <what>`, so if you are changing stuff in libyelp directory the `project part` could be `libyelp`, if in `src`, then `yelp`, and so on.

Hints for commit messages:
- unref - It was basically about keeping parent URI alive when resolving the base URI (in our case - base URI was the one with `YELP_GNOME_HELP_URI` and parent one with added `xref:stuff` or else.
- YELP_SKIP_GUI - you can mention that this may be needed for projects that want to use libyelp, but have no GUI, so they may want to skip some GUI parts libyelp is trying to initialize and aborts because of some failure. Mention that one of such projects will be search provider for yelp.
- YELP_GNOME_HELP_URI - so other projects can use libyelp and avoid copying the `help:gnome-help` string. Search provider will use it.
- actions in Yelp - maybe explain shortly what actions did you add ("show-page" for showing a specific page and "show-search" for showing a search page for a given term), will be used by search provider.
- search provider - probably `search-provider: New`
